File: packages/python/palpluss/http/transport.py
from __future__ import annotations

import base64
import time
from typing import Any

import httpx

from .errors import PalPlussApiError, RateLimitError

DEFAULT_BASE_URL = "https://api.palpluss.com/v1"
DEFAULT_TIMEOUT = 30.0
DEFAULT_MAX_RETRIES = 3


def _build_auth_header(api_key: str) -> str:
    return "Basic " + base64.b64encode(f"{api_key}:".encode()).decode()


def _build_url(base_url: str, path: str) -> str:
    base = base_url.rstrip("/")
    return base + (path if path.startswith("/") else "/" + path)


def _clean_query(query: dict[str, Any] | None) -> dict[str, str] | None:
    if not query:
        return None
    return {k: str(v) for k, v in query.items() if v is not None}


def _handle_response(response: httpx.Response) -> Any:
    """Unwrap the API envelope or raise a typed error."""
    if response.status_code == 204:
        return None

    body: dict[str, Any] = response.json()
    request_id: str | None = body.get("requestId")

    if not response.is_success or body.get("success") is False:
        error = body.get("error") or {}
        api_error = PalPlussApiError.from_response(response.status_code, error, request_id)
        if isinstance(api_error, RateLimitError):
            retry_after_header = response.headers.get("Retry-After")
            if retry_after_header:
                api_error.retry_after = int(retry_after_header)
        raise api_error

    return body.get("data")


class HttpTransport:
    """Synchronous HTTP transport layer."""

    def __init__(
        self,
        api_key: str,
        base_url: str = DEFAULT_BASE_URL,
        timeout: float = DEFAULT_TIMEOUT,
        auto_retry_on_rate_limit: bool = True,
        max_retries: int = DEFAULT_MAX_RETRIES,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._auto_retry_on_rate_limit = auto_retry_on_rate_limit
        self._max_retries = max_retries
        self._client = httpx.Client(
            headers={
                "Authorization": _build_auth_header(api_key),
                "Accept": "application/json",
            },
            timeout=timeout,
        )

    def request(
        self,
        method: str,
        path: str,
        *,
        body: dict[str, Any] | None = None,
        query: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> Any:
        url = _build_url(self._base_url, path)
        extra_headers: dict[str, str] = {}
        if idempotency_key:
            extra_headers["Idempotency-Key"] = idempotency_key

        send_body = body if body and method in ("POST", "PATCH", "PUT") else None
        params = _clean_query(query)

        attempt = 0
        while True:
            response = self._client.request(
                method,
                url,
                headers=extra_headers or None,
                json=send_body,
                params=params,
            )

            rate_limited = response.status_code == 429
            if rate_limited and self._auto_retry_on_rate_limit and attempt < self._max_retries:
                retry_after = int(response.headers.get("Retry-After", "1"))
                time.sleep(retry_after)
                attempt += 1
                continue

            return _handle_response(response)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> HttpTransport:
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()


class AsyncHttpTransport:
    """Asynchronous HTTP transport layer."""

    def __init__(
        self,
        api_key: str,
        base_url: str = DEFAULT_BASE_URL,
        timeout: float = DEFAULT_TIMEOUT,
        auto_retry_on_rate_limit: bool = True,
        max_retries: int = DEFAULT_MAX_RETRIES,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._auto_retry_on_rate_limit = auto_retry_on_rate_limit
        self._max_retries = max_retries
        self._client = httpx.AsyncClient(
            headers={
                "Authorization": _build_auth_header(api_key),
                "Accept": "application/json",
            },
            timeout=timeout,
        )

    async def request(
        self,
        method: str,
        path: str,
        *,
        body: dict[str, Any] | None = None,
        query: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> Any:
        import asyncio

        url = _build_url(self._base_url, path)
        extra_headers: dict[str, str] = {}
        if idempotency_key:
            extra_headers["Idempotency-Key"] = idempotency_key

        send_body = body if body and method in ("POST", "PATCH", "PUT") else None
        params = _clean_query(query)

        attempt = 0
        while True:
            response = await self._client.request(
                method,
                url,
                headers=extra_headers or None,
                json=send_body,
                params=params,
            )

            rate_limited = response.status_code == 429
            if rate_limited and self._auto_retry_on_rate_limit and attempt < self._max_retries:
                retry_after = int(response.headers.get("Retry-After", "1"))
                await asyncio.sleep(retry_after)
                attempt += 1
                continue

            return _handle_response(response)

    async def close(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> AsyncHttpTransport:
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.close()

File: packages/go/transport.go
package palpluss

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

type transport struct {
	client     *http.Client
	baseURL    string
	authHeader string
	autoRetry  bool
	maxRetries int
}

func newTransport(apiKey, baseURL string, timeout time.Duration, autoRetry bool, maxRetries int) *transport {
	return &transport{
		client:     &http.Client{Timeout: timeout},
		baseURL:    strings.TrimRight(baseURL, "/"),
		authHeader: "Basic " + base64.StdEncoding.EncodeToString([]byte(apiKey+":")),
		autoRetry:  autoRetry,
		maxRetries: maxRetries,
	}
}

type apiEnvelope struct {
	Success   bool            `json:"success"`
	RequestID string          `json:"requestId"`
	Data      json.RawMessage `json:"data"`
	Error     struct {
		Message string         `json:"message"`
		Code    string         `json:"code"`
		Details map[string]any `json:"details"`
	} `json:"error"`
}

func (t *transport) do(
	ctx context.Context,
	method, path string,
	body map[string]any,
	query url.Values,
	idempotencyKey string,
) (json.RawMessage, error) {
	rawURL := t.baseURL
	if strings.HasPrefix(path, "/") {
		rawURL += path
	} else {
		rawURL += "/" + path
	}
	if len(query) > 0 {
		rawURL += "?" + query.Encode()
	}

	sendBody := body != nil && (method == "POST" || method == "PATCH" || method == "PUT")
	var bodyBytes []byte
	if sendBody {
		var err error
		bodyBytes, err = json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("palpluss: marshal request body: %w", err)
		}
	}

	for attempt := 0; ; attempt++ {
		var bodyReader io.Reader
		if bodyBytes != nil {
			bodyReader = bytes.NewReader(bodyBytes)
		}

		req, err := http.NewRequestWithContext(ctx, method, rawURL, bodyReader)
		if err != nil {
			return nil, fmt.Errorf("palpluss: build request: %w", err)
		}

		req.Header.Set("Authorization", t.authHeader)
		req.Header.Set("Accept", "application/json")
		if bodyBytes != nil {
			req.Header.Set("Content-Type", "application/json")
		}
		if idempotencyKey != "" {
			req.Header.Set("Idempotency-Key", idempotencyKey)
		}

		resp, err := t.client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("palpluss: http: %w", err)
		}

		if resp.StatusCode == http.StatusTooManyRequests && t.autoRetry && attempt < t.maxRetries {
			retryAfter := 1
			if v := resp.Header.Get("Retry-After"); v != "" {
				if n, err2 := strconv.Atoi(v); err2 == nil && n > 0 {
					retryAfter = n
				}
			}
			resp.Body.Close()
			time.Sleep(time.Duration(retryAfter) * time.Second)
			continue
		}

		return handleResponse(resp)
	}
}

func handleResponse(resp *http.Response) (json.RawMessage, error) {
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNoContent {
		return nil, nil
	}

	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("palpluss: read response body: %w", err)
	}

	var env apiEnvelope
	if err := json.Unmarshal(b, &env); err != nil {
		return nil, fmt.Errorf("palpluss: decode response: %w", err)
	}

	isSuccess := resp.StatusCode >= 200 && resp.StatusCode < 300
	if !isSuccess || !env.Success {
		msg := env.Error.Message
		if msg == "" {
			msg = "Unknown error"
		}
		code := env.Error.Code
		if code == "" {
			code = "UNKNOWN"
		}
		apiErr := &APIError{
			Message:    msg,
			ErrorCode:  code,
			HTTPStatus: resp.StatusCode,
			Details:    env.Error.Details,
			RequestID:  env.RequestID,
		}
		if resp.StatusCode == http.StatusTooManyRequests {
			rl := &RateLimitError{APIError: apiErr}
			if v := resp.Header.Get("Retry-After"); v != "" {
				if n, err2 := strconv.Atoi(v); err2 == nil {
					rl.RetryAfter = n
				}
			}
			return nil, rl
		}
		return nil, apiErr
	}

	return env.Data, nil
}

// call executes a request and decodes the data payload into T.
func call[T any](
	ctx context.Context,
	t *transport,
	method, path string,
	body map[string]any,
	query url.Values,
	idempotencyKey string,
) (*T, error) {
	raw, err := t.do(ctx, method, path, body, query, idempotencyKey)
	if err != nil {
		return nil, err
	}
	if raw == nil {
		return nil, nil
	}
	var result T
	if err := json.Unmarshal(raw, &result); err != nil {
		return nil, fmt.Errorf("palpluss: decode data: %w", err)
	}
	return &result, nil
}
